From 346493df139eafedb91872ac39df788dcecb0979 Mon Sep 17 00:00:00 2001
From: Jayasankar <jayasankarx.nara@intel.com>
Date: Tue, 15 Nov 2016 17:24:10 +0800
Subject: [PATCH] igb: Do not disconnect netdevice from system during runtime
 suspend

During runtime suspend in __igb_shutdown() function do not
mark device as removed from system, in igb_down() function
do not indicate loss of carrier and do not stop upper layers
from calling transmit function.

Signed-off-by: Jayasankar <jayasankarx.nara@intel.com>
Signed-off-by: Lai, Poey Seng <poey.seng.lai@intel.com>
Reviewed-by: Vinod Kumar <vinod.kumarx.vinod.kumar@intel.com>
---
 drivers/net/ethernet/intel/igb/igb_main.c | 31 +++++++++++++++++++++++++------
 1 file changed, 25 insertions(+), 6 deletions(-)

diff --git a/drivers/net/ethernet/intel/igb/igb_main.c b/drivers/net/ethernet/intel/igb/igb_main.c
index 1f6f424..28a2da6 100644
--- a/drivers/net/ethernet/intel/igb/igb_main.c
+++ b/drivers/net/ethernet/intel/igb/igb_main.c
@@ -2166,11 +2166,15 @@ void igb_down(struct igb_adapter *adapter)
 	struct e1000_hw *hw = &adapter->hw;
 	u32 tctl, rctl;
 	int i;
-
+	struct pci_dev *pdev = adapter->pdev;
 	/* signal that we're down so the interrupt handler does not
 	 * reschedule our watchdog timer
 	 */
-	set_bit(__IGB_DOWN, &adapter->state);
+	/* While in suspend state the i210 FW is not doing Link Down
+	 * so no need to set state to __IGB_DOWN
+	 */
+	if (!(pdev->dev.power.runtime_status == RPM_SUSPENDING))
+		set_bit(__IGB_DOWN, &adapter->state);
 
 	/* disable receives in the hardware */
 	rctl = rd32(E1000_RCTL);
@@ -2179,8 +2183,14 @@ void igb_down(struct igb_adapter *adapter)
 
 	igb_nfc_filter_exit(adapter);
 
-	netif_carrier_off(netdev);
-	netif_tx_stop_all_queues(netdev);
+	/* for supporting runtime pm based on activity monitoring
+	 * do not stop the queue so that apps can continue using
+	 * the net device as they are not aware of runtime suspend
+	 */
+	if (!(pdev->dev.power.runtime_status == RPM_SUSPENDING)) {
+		netif_carrier_off(netdev);
+		netif_tx_stop_all_queues(netdev);
+	}
 
 	/* disable transmits in the hardware */
 	tctl = rd32(E1000_TCTL);
@@ -2192,7 +2202,11 @@ void igb_down(struct igb_adapter *adapter)
 
 	igb_irq_disable(adapter);
 
-	adapter->flags &= ~IGB_FLAG_NEED_LINK_UPDATE;
+	/* While in suspend state the i210 FW is not doing Link Down
+	 * so no need to clear flag IGB_FLAG_NEED_LINK_UPDATE
+	 */
+	if (!(pdev->dev.power.runtime_status == RPM_SUSPENDING))
+		adapter->flags &= ~IGB_FLAG_NEED_LINK_UPDATE;
 
 	for (i = 0; i < adapter->num_q_vectors; i++) {
 		if (adapter->q_vector[i]) {
@@ -8779,7 +8793,12 @@ static int __igb_shutdown(struct pci_dev *pdev, bool *enable_wake,
 	bool wake;
 
 	rtnl_lock();
-	netif_device_detach(netdev);
+	/* For supporting runtime pm based on activity monitoring
+	 * do not disconnect net device from stack while entering
+	 * runtime suspend
+	 */
+	if (!runtime)
+		netif_device_detach(netdev);
 
 	if (netif_running(netdev))
 		__igb_close(netdev, true);
-- 
1.9.1

